openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/changeEvent/AbstractEventConsumer.java

```diff
@@ -0,0 +1,234 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.openmetadata.schema.entity.events.EventSubscription;
+import org.openmetadata.schema.entity.events.EventSubscriptionOffset;
+import org.openmetadata.schema.entity.events.SubscriptionStatus;
+import org.openmetadata.schema.type.ChangeEvent;
+import org.openmetadata.service.Entity;
+import org.openmetadata.service.events.errors.RetriableException;
+import org.openmetadata.service.events.subscription.AlertUtil;
+import org.openmetadata.service.util.JsonUtils;
+import org.quartz.DisallowConcurrentExecution;
+import org.quartz.Job;
+import org.quartz.JobDetail;
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+
+@Slf4j
+@DisallowConcurrentExecution
+public abstract class AbstractEventConsumer implements Consumer<ChangeEvent>, Job {
+  public static final String ALERT_OFFSET_KEY = "alertOffsetKey";
+  public static final String ALERT_INFO_KEY = "alertInfoKey";
+  private static final String OFFSET_EXTENSION = "eventSubscription.Offset";
+  protected static final int BACKOFF_NORMAL = 0;
+  protected static final int BACKOFF_3_SECONDS = 3 * 1000;
+  protected static final int BACKOFF_30_SECONDS = 30 * 1000;
+  protected static final int BACKOFF_5_MINUTES = 5 * 60 * 1000;
+  protected static final int BACKOFF_1_HOUR = 60 * 60 * 1000;
+  protected static final int BACKOFF_24_HOUR = 24 * 60 * 60 * 1000;
+
+  @Getter protected int currentBackoffTime = BACKOFF_NORMAL;
+  private int offset = -1;
+
+  @Getter @Setter private JobDetail jobDetail;
+  protected EventSubscription eventSubscription;
+
+  protected AbstractEventConsumer() {}
+
+  private void init(JobExecutionContext context) {
+    EventSubscription sub = (EventSubscription) context.getJobDetail().getJobDataMap().get(ALERT_INFO_KEY);
+    this.jobDetail = context.getJobDetail();
+    this.eventSubscription = sub;
+    this.offset = loadInitialOffset();
+    this.doInit(context);
+  }
+
+  protected abstract void doInit(JobExecutionContext context);
+
+  protected void sendAlert(List<ChangeEvent> list) {
+    /* This method needs to be over-ridden by specific Publisher for sending Alert */
+
+  }
+
+  @Override
+  public void handleFailedEvents(List<ChangeEvent> failedEvents) {}
+
+  @Override
+  public void handleException(Exception e) {}
+
+  private int loadInitialOffset() {
+    int eventSubscriptionOffset;
+    String json =
+        Entity.getCollectionDAO()
+            .eventSubscriptionDAO()
+            .getSubscriberOffset(eventSubscription.getId().toString(), OFFSET_EXTENSION);
+    if (json != null) {
+      EventSubscriptionOffset offsetFromDb = JsonUtils.readValue(json, EventSubscriptionOffset.class);
+      eventSubscriptionOffset = offsetFromDb.getOffset();
+    } else {
+      eventSubscriptionOffset = Entity.getCollectionDAO().changeEventDAO().listCount();
```

Comment:
what does the listCount() do here, shouldn't the offset be the earliest or latest offset in change_event_table?

Reply:
the listCount is the latest count of the change_event table